streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/HighAvailabilityTaskAssignor.java

```diff
@@ -0,0 +1,544 @@
+    }
+
+    /**
+     * Rankings are computed as follows, with lower being more caught up:
+     *      Rank -1: active running task
+     *      Rank 0: standby or restoring task whose overall lag is within the acceptableRecoveryLag bounds
+     *      Rank 1: tasks whose lag is unknown, eg because it was not encoded in an older version subscription
+     *      Rank 1+: all other tasks are ranked according to their actual total lag
+     * @return Sorted set of all client candidates for each stateful task, ranked by their overall lag. Tasks are
+     */
+    static <ID extends Comparable<ID>> SortedMap<TaskId, SortedSet<RankedClient<ID>>> buildClientRankingsByTask(final Set<TaskId> statefulTasks,
+                                                                                                                final Map<ID, ClientState> clientStates,
+                                                                                                                final long acceptableRecoveryLag) {
+        final SortedMap<TaskId, SortedSet<RankedClient<ID>>> statefulTasksToRankedCandidates = new TreeMap<>();
+
+        for (final TaskId task : statefulTasks) {
+            final SortedSet<RankedClient<ID>> rankedClientCandidates = new TreeSet<>();
+            statefulTasksToRankedCandidates.put(task, rankedClientCandidates);
+
+            for (final Map.Entry<ID, ClientState> clientEntry : clientStates.entrySet()) {
+                final ID clientId = clientEntry.getKey();
+                final long taskLag = clientEntry.getValue().lagFor(task);
+                final long clientRank;
+                if (taskLag == Task.LATEST_OFFSET) {
+                    clientRank = Task.LATEST_OFFSET;
+                } else if (taskLag == UNKNOWN_OFFSET_SUM) {
+                    clientRank = 1L;
+                } else if (taskLag <= acceptableRecoveryLag) {
+                    clientRank = 0L;
+                } else {
+                    clientRank = taskLag;
+                }
+                rankedClientCandidates.add(new RankedClient<>(clientId, clientRank));
+            }
+        }
+        log.trace("Computed statefulTasksToRankedCandidates map as {}", statefulTasksToRankedCandidates);
+
+        return statefulTasksToRankedCandidates;
+    }
+
+    /**
+     * Compute the balance factor as the difference in stateful active task count per thread between the most and
+     * least loaded clients
+     */
+    static int computeBalanceFactor(final Collection<ClientState> clientStates,
+                                    final Set<TaskId> statefulTasks,
+                                    final Function<ClientState, Set<TaskId>> activeTaskSet) {
+        int minActiveStatefulTasksPerThreadCount = Integer.MAX_VALUE;
+        int maxActiveStatefulTasksPerThreadCount = 0;
+
+        for (final ClientState state : clientStates) {
+            final Set<TaskId> activeTasks = new HashSet<>(activeTaskSet.apply(state));
```

Comment:
I feel like I might be missing something, but what's the advantage of creating a method reference and passing it in so that we can invoke it with the object that has that method? I.e., why not just:
```suggestion
            final Set<TaskId> activeTasks = new HashSet<>(state.prevActiveTasks());
```
And then we don't need the third parameter?

Reply:
Refactoring artifact...originally we would also compute the balance factor for the final assignment for some optimization, but I took it out to keep things simple (relatively speaking...). So now this makes no sense, I agree